common/src/main/java/oracle/kubernetes/common/utils/SchemaConversionUtils.java

```diff
@@ -104,6 +120,128 @@ public String convertDomainSchema(String domainYaml) {
     return yaml.dump(convertDomainSchema((Map<String, Object>) domain));
   }
 
+  private void convertClustersToResources(ClusterCustomResourceHelper clusterCustomResourceHelper,
+      Map<String, Object> domain, String desiredAPIVersion) {
+    try {
+      if (clusterCustomResourceHelper != null) {
+        String apiVersion = (String)domain.get("apiVersion");
+        if (CommonConstants.API_VERSION_V8.equals(apiVersion)
+            && CommonConstants.API_VERSION_V9.equals(desiredAPIVersion)) {
+          convertV8toV9(clusterCustomResourceHelper, domain);
+        } else if (CommonConstants.API_VERSION_V9.equals(apiVersion)
+            && CommonConstants.API_VERSION_V8.equals(desiredAPIVersion)) {
+          convertV9toV8(clusterCustomResourceHelper, domain);
+        }
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  protected void convertV8toV9(ClusterCustomResourceHelper clusterCustomResourceHelper, Map<String, Object> domain)
+      throws ApiException {
+    Map<String, Object> domainSpec = getSpec(domain);
+    List<Object> clusterSpecs = getOrCreateClusterSpecs(domainSpec);
+    if (!clusterSpecs.isEmpty()) {
+      List<String> clusterResources = clusterCustomResourceHelper
+          .getNamesOfDeployedClusterResources(getNamespace(domain), getDomainUid(domain));
+      Iterator<Object> itr = clusterSpecs.iterator();
+      while (itr.hasNext()) {
+        Map<String, Object> clusterSpec = (Map<String, Object>) itr.next();
+        String clusterName = (String) clusterSpec.get("clusterName");
+        if (!clusterResources.contains(clusterName)) {
+          // Create Cluster Resource
+          clusterCustomResourceHelper.createClusterResource(clusterSpec, domain);
+
+          // remove the cluster spec from the map
+          itr.remove();
+
+          // Add the reference to the DomainSpec
+          addClusterResourceReference(domainSpec, clusterName);
+        }
+      }
+    }
+  }
+
+  private void addClusterResourceReference(Map<String, Object> domainSpec, String clusterName) {
+    Optional.ofNullable(getClusterResourceReferences(domainSpec)).map(c -> c.add(clusterName));
+  }
+
+  protected List<String> getClusterResourceReferences(Map<String, Object> domainSpec) {
+    List<String> clusterResourceReferences =
+        (List<String>) domainSpec.get(CLUSTER_RESOURCE_REFERENCES);
```

Comment:
"cluster-resource-references" is a mouthful... I'm wondering if we can just still use "clusters" but have the content now be the list of references thereby highlighting that the schema is mostly unchanged but just that this content has been promoted to its own resource.

Reply:
Yes, I that should be able to be done.